### PR TITLE
add protobuf and use that to format the waf log in json format

### PR DESCRIPTION
**Pull request by azhao155**
Integrate protobuf into modsec  and use that to print out modsec fields into json format to a file.
for detail: please check 
https://microsoft.sharepoint.com/teams/WAG/AzureNetworking/_layouts/OneNote.aspx?id=%2Fteams%2FWAG%2FAzureNetworking%2FEngineering%20H%20%20P%2FHybrid%2FGENERAL%2FOneNote%2FBrooklyn%20Engineering%20Handbook&wd=target%28AppGW%20Linux%20VMSS.one%7C9DC62AD6-5B8D-4256-9317-C10C543FA908%2FIntegrate%20Protobuf%20into%20Modsec%7C0F0E4867-591C-47AA-9CFD-3112A1DABA61%2F%29
onenote:https://microsoft.sharepoint.com/teams/WAG/AzureNetworking/Engineering%20H%20%20P/Hybrid/GENERAL/OneNote/Brooklyn%20Engineering%20Handbook/AppGW%20Linux%20VMSS.one#Integrate%20Protobuf%20into%20Modsec&section-id={9DC62AD6-5B8D-4256-9317-C10C543FA908}&page-id={0F0E4867-591C-47AA-9CFD-3112A1DABA61}&end

this is for review. and following changes needs to be done.
1. Improve the format, hide the full path of rule file and escape characters
2. make log filename configurable
3. add time stamp

Sample logging format:

original:

2018/06/03 05:46:36 [error] 127559#0: [client 167.220.24.99] ModSecurity: Warning. Pattern match "^[\\d.:]+$" at REQUEST_HEADERS:Host. [file "/root/CP/Nginx/crs/crs-3.0/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "810"] [id "920350"] [rev "2"] [msg "Host header is a numeric IP address"] [data "52.191.172.77:8080"] [severity "WARNING"] [ver "OWASP_CRS/3.0.0"] [maturity "9"] [accuracy "9"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS/PROTOCOL_VIOLATION/IP_HOST"] [tag "WASCTC/WASC-21"] [tag "OWASP_TOP_10/A7"] [tag "PCI/6.5.10"] [hostname ""] [uri "/icons/ubuntu-logo.png"] [unique_id "0cBGhchcAcAcAcAcecnyAcA0"]

{
 "resourceId": "",
 "operationName": "ApplicationGatewayFirewall",
 "time": "2018-06-03T05:46:36Z",
 "category": "ApplicationGatewayFirewallLog",
 "properties": {
  "instanceId": "",
  "clientIp": "167.220.24.99",
  "clientPort": "",
  "requestUri": "/icons/ubuntu-logo.png",
  "ruleSetType": "OWASP_CRS",
  "ruleSetVersion": "3.0.0",
  "ruleId": "\"920350\"",
  "message": "\"Host header is a numeric IP address\"",
  "action": "Detected",
  "site": "Global",
  "details": {
   "message": "Warning. Pattern match \"^[\\\\d.:]+$\" at REQUEST_HEADERS:Host. ",
   "data": "\"52.191.172.77:8080\"",
   "file": "\"rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf\"",
   "line": "\"810\""
  },
  "hostname": ""
 }
}

### CODE REVIEW

File: apache2/waf_logging/waf_log_util_internal.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef _WAF_LOG_UTIL_INTERNAL_HEADER
+#define _WAF_LOG_UTIL_INTERNAL_HEADER
+
+#include <google/protobuf/util/json_util.h>
+#include"waf_format.pb.h"
+#include <fstream>
+
+#define WAF_LOG_PATH "/var/log/Appgwwaf/waf_json.log"
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util_internal.h]: WAF_LOG_PATH is a little bit general

File: nginx/modsecurity/config.in [hunk 2]
```diff
@@ -22,7 +22,9 @@ CORE_LIBS="$CORE_LIBS \
     @PCRE_LDADD@ \
     @APXS_LIBS@ \
     @YAJL_LIBS@ \
-    @SSDEEP_LDFLAGS@"
+    @SSDEEP_LDFLAGS@ \
+    -lstdc++ \
```
Comment (cp5555) [inline nginx/modsecurity/config.in:26]: Do we need to change to @xx@

File: apache2/waf_logging/waf_log_util.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
+#include "waf_log_util_internal.h"
+#include "waf_log_util_external.h"
+
+// This function fills in a Person message based on user input.
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: could we improve the comments?

File: apache2/waf_logging/waf_log_util.cc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  if (messages != NULL) {
+      waf_format->set_messages(messages);
+  }
+
+  switch(action) {
+      case 1:
+          waf_format->set_action(waf_format::Waf_Format::Detected);
+      case 2:
+          waf_format->set_action(waf_format::Waf_Format::Blocked);
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Could we add default?

File: apache2/waf_logging/waf_log_util.cc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  if (site == 0) {
+      waf_format->set_site(waf_format::Waf_Format::Global);
+  }
+
+  waf_format::Details details;
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Could we move to the beggining of this function?

File: apache2/waf_logging/waf_log_util.cc [hunk 6]
```diff
@@ -0,0 +1,104 @@
+#include "waf_log_util_internal.h"
+#include "waf_log_util_external.h"
+
+// This function fills in a Person message based on user input.
+void set_waf_format(waf_format::Waf_Format* waf_format, char* clientIP, char* clientPort, char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, char* hostname) {
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Do we need to do some failure check in this function?

File: apache2/waf_logging/waf_log_util.cc [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  }
+}
+
+// Main function:  Reads the entire address book from a file,
+//   adds one person based on user input, then writes it back out to the same
+//   file.
+int write_json_to_file(char* clientIP, char* clientPort, char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, char* hostname) {
+  // Verify that the version of the library that we linked against is
+  // compatible with the version of the headers we compiled against.
+  GOOGLE_PROTOBUF_VERIFY_VERSION;
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Do we need to do some failure handling?

File: apache2/waf_logging/waf_log_util.cc [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  std::ofstream json_file;
+  std::string json_string;
+  json_file.open(WAF_LOG_PATH, std::ios::app);
+  if (json_file.fail())
+  {
+     return WAF_LOG_OPEN_FAILED;
+  }
+
+  waf_format::Waf_Format waf_format;
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Could we move the funtion declaration to the beginning?

File: apache2/waf_logging/waf_log_util.cc [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  MessageToJsonString(waf_format, &json_string, options);
+
+    // Write the new address book back to disk.
+  json_file << json_string << std::endl;
+
+  json_file.close();
+  // Optional:  Delete all global objects allocated by libprotobuf.
+  //google::protobuf::ShutdownProtobufLibrary();
+
+  return 0;
```
Comment (cp5555) [inline apache2/waf_logging/waf_log_util.cc]: Do we need a marco for the success?

File: apache2/apache2_util.c [hunk 10]
```diff
@@ -187,6 +189,83 @@ char *get_env_var(request_rec *r, char *name) {
...
+        strncpy(output, first, last - first);
+    }
+}
+
+/**
+ * send all waf fields in json format to a file.
+ */
+void send_waf_log(const char* str1, const char* ip_port, const char* uri, int mode, const char* hostname, request_rec *r) {
+    int rc = 0;
+    char waf_filename[1024] = "";
```
Comment (cp5555) [inline apache2/apache2_util.c]: Do we need to add macro for the filename length and ip length?

File: apache2/apache2_util.c [hunk 11]
```diff
@@ -187,6 +189,83 @@ char *get_env_var(request_rec *r, char *name) {
...
+    else {
+        strcpy(waf_ip, ip_port);
+    }
+
+    char *end = strstr(str1, "[file ");
+    if (end != NULL) {
+        strncpy(waf_detail_message, str1, end - str1);
+    }
+
+    rc = write_json_to_file(waf_ip, waf_port, uri, "", "", waf_id, waf_message, mode, 0, waf_detail_message, waf_data, waf_filename, waf_line, hostname);
```
Comment (cp5555) [inline apache2/apache2_util.c]: "" or NULL?

File: apache2/waf_logging/waf_log_util_internal.h (hunk 1 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util_internal.h]: make it configurable now

File: apache2/waf_logging/waf_log_util.cc (hunk 5 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: done

File: apache2/waf_logging/waf_log_util.cc (hunk 4 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: done

File: apache2/waf_logging/waf_log_util.cc (hunk 6 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: put in the try catch

File: apache2/waf_logging/waf_log_util.cc (hunk 7 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: put in the try catch

File: apache2/waf_logging/waf_log_util.cc (hunk 8 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: done

File: apache2/waf_logging/waf_log_util.cc (hunk 9 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: added

File: apache2/apache2_util.c (hunk 11 above)
Comment (azhao155) [reply apache2/apache2_util.c]: will add directory to set rule set and version

File: apache2/mod_security2.c [hunk 12]
```diff
@@ -92,6 +92,11 @@ unsigned long int DSOLOCAL conn_write_state_limit = 0;
 TreeRoot DSOLOCAL *conn_write_state_whitelist = 0;
 TreeRoot DSOLOCAL *conn_write_state_suspicious_list = 0;
 
+#ifdef WAF_JSON_LOGGING_ENABLE
```
Comment (meetug) [inline apache2/mod_security2.c:95]: I am not sure if we need this change. I looked at sample audit logs and there is already a ver variable that has the rule set type and version information (e.g. ver "OWASP_CRS/3.0.0"). So, somehow that information is already getting plumbed?

File: apache2/apache2_util.c [hunk 13]
```diff
@@ -187,6 +192,115 @@ char *get_env_var(request_rec *r, char *name) {
...
+#ifdef WAF_JSON_LOGGING_ENABLE
+/**
+ * Retrieve waf log field.
+ */
+void get_field_value(const char* from, const char* to, const char* text, char* output) {
+    char* first = strstr(text, from);
+    int first_index = first - text;
+
+    if (first != NULL ) {
+        if ((first_index > 0 && (first-1)[0] != "\\") || (first_index == 0)) {
```
Comment (allanbomsft) [inline apache2/apache2_util.c:204]: be 100% sure that it works ] in msg, or \", or \\

File: apache2/apache2_util.c [hunk 14]
```diff
@@ -187,6 +192,115 @@ char *get_env_var(request_rec *r, char *name) {
...
+        }
+    }
+}
+
+/**
+ * send all waf fields in json format to a file.
+ */
+void send_waf_log(const char* data_dir, const char* str1, const char* ip_port, const char* uri, const char* time, int mode, const char* hostname, request_rec *r) {
+    int rc = 0;
+    char waf_filename[1024] = "";
```
Comment (allanbomsft) [inline apache2/apache2_util.c]: what happens if filename is > 1024?

File: apache2/waf_logging/waf_log_util.cc [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    
+    if (data_dir == NULL) {
+       return WAF_LOG_UTIL_FAILED;
+    }
+    
+    directory_name = data_dir;
+    file_name = WAF_LOG_UTIL_FILE;
+    fullname = (directory_name + "/" + file_name).c_str();
+
+    json_file.open(fullname, std::ios::app);
```
Comment (allanbomsft) [inline apache2/waf_logging/waf_log_util.cc]: should we perhaps open this only once, and not on every single write?

File: apache2/waf_logging/waf_log_util.cc [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    std::string file_name;
+    const char* fullname;
+    
+    if (data_dir == NULL) {
+       return WAF_LOG_UTIL_FAILED;
+    }
+    
+    directory_name = data_dir;
+    file_name = WAF_LOG_UTIL_FILE;
+    fullname = (directory_name + "/" + file_name).c_str();
```
Comment (allanbomsft) [inline apache2/waf_logging/waf_log_util.cc]: is this file guarenteed to be rotated?

File: apache2/apache2_util.c (hunk 14 above)
Comment (azhao155) [reply apache2/apache2_util.c]: the maxium of str1 is 1024, so won't exceed this value

File: apache2/waf_logging/waf_log_util.cc (hunk 15 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc]: yes, made the change

File: apache2/mod_security2.c (hunk 12 above)
Comment (azhao155) [reply apache2/mod_security2.c:95]: yes, remove it.

**Review by meetug — APPROVED** [review]

**Review by meetug — COMMENTED** [review]
As discussed, we need to make changes to add ResourceId, InstanceId fields.

File: apache2/waf_logging/waf_format.proto [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+message Properties {
+  optional string instanceId = 1;
+  optional string clientIP = 2;
+  optional string clientPort = 3;
+  optional string requestUri = 4;
+  optional string ruleSetType = 5;
+  optional string ruleSetVersion = 6;
+  optional string ruleId = 7;
+  optional string messages = 8;
```
Comment (meetug) [inline apache2/waf_logging/waf_format.proto]: Should this be message?

File: apache2/waf_logging/waf_format.proto [hunk 18]
```diff
@@ -0,0 +1,43 @@
+syntax = "proto2";
+
+package waf_format;
+
+message Details {
+  optional string messages = 1;
```
Comment (meetug) [inline apache2/waf_logging/waf_format.proto]: Same here, should this be message?

File: apache2/apache2_config.c [hunk 19]
```diff
@@ -3958,6 +4016,22 @@ const command_rec module_directives[] = {
...
+        NULL,
+        CMD_SCOPE_ANY,
+        "Set waf resourceId"
+    ),
+    AP_INIT_TAKE1 (
+        "SecWafInstanceId",
+        cmd_waf_instanceId,
+        NULL,
+        CMD_SCOPE_ANY,
+        "Set waf instancdId"
```
Comment (meetug) [inline apache2/apache2_config.c]: Minor, typo: instanceId

File: apache2/waf_logging/waf_log_util_external.h [hunk 20]
```diff
@@ -0,0 +1,20 @@
+#ifndef _WAF_LOG_UTIL_EXTERNAL_HEADER
+#define _WAF_LOG_UTIL_EXTERNAL_HEADER
+
+#define WAF_LOG_UTIL_FAILED -1
+#define WAF_LOG_UTIL_SUCCESS 0
+#define WAF_RULESET_PREFIX "/RuleSets/"
+#define WAF_LOG_UTIL_FILE "waf_json.log"
+#define WAF_LOG_UTIL_OPERATION_NAME "ApplicationGatewayFirewall"
```
Comment (meetug) [inline apache2/waf_logging/waf_log_util_external.h:8]: Should we add these as defaults in the proto definition?

File: apache2/waf_logging/waf_format.proto (hunk 17 above)
Comment (azhao155) [reply apache2/waf_logging/waf_format.proto]: will fix it

File: apache2/waf_logging/waf_format.proto (hunk 18 above)
Comment (azhao155) [reply apache2/waf_logging/waf_format.proto]: will fix it

File: apache2/apache2_config.c (hunk 19 above)
Comment (azhao155) [reply apache2/apache2_config.c]: will fix it

File: apache2/waf_logging/waf_log_util_external.h (hunk 20 above)
Comment (azhao155) [reply apache2/waf_logging/waf_log_util_external.h:8]: i think it's better to pull all the waf related dependency out of protobuf.  Protobuf should be a standalone utility for us and we put all the waf related info in the header file we exposed.

File: apache2/apache2_util.c (hunk 13 above)
Comment (azhao155) [reply apache2/apache2_util.c:204]: yes, checked it

File: apache2/waf_logging/waf_format.proto [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+  optional string data = 2;
+  optional string file = 3;
+  optional string line = 4;
+}
+
+
+message Properties {
+  optional string instanceId = 1;
+  optional string clientIP = 2;
```
Comment (meetug) [inline apache2/waf_logging/waf_format.proto]: should be clientIp

Comment (azhao155) [reply apache2/waf_logging/waf_format.proto]: fixed it in the next commit

File: apache2/waf_logging/waf_log_util.cc [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+    waf_format::Waf_Format waf_format;
+    std::string json_string;
+    google::protobuf::util::JsonPrintOptions options;
+    google::protobuf::util::Status convert_result;
+    char* json_str;
+    
+    try {
+        // Verify that the version of the library that we linked against is
+        // compatible with the version of the headers we compiled against.
+        GOOGLE_PROTOBUF_VERIFY_VERSION;
```
Comment (meetug) [inline apache2/waf_logging/waf_log_util.cc:107]: Should we do this just once during initialization?

Comment (azhao155) [reply apache2/waf_logging/waf_log_util.cc:107]: I think so, but i don't want to expose this to the modsec or add wrapper to make it complicated, also for the performance wise, it doesn't degrade.. so i think it should be fine

**Review by meetug — APPROVED** [review]

File: apache2/apache2_util.c (hunk 13 above)
Comment (asridharan) [reply apache2/apache2_util.c:204]: @azhao155 could you make this function `static`

File: apache2/apache2_util.c [hunk 23]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+
+    if (first != NULL && last != NULL) {
+        strncpy(output, first, last - first);
+    }
+}
+
+/**
+ * get ip and port number.
+ */
+void get_ip_port(const char* ip_port, char* waf_ip, char* waf_port) {
```
Comment (asridharan) [inline apache2/apache2_util.c:233]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 24]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+    }
+    else {
+        strcpy(waf_ip, ip_port);
+    }
+}
+
+/**
+ * get detail_messages.
+ */
+void get_detail_message(const char* str1, char* waf_detail_message) {
```
Comment (asridharan) [inline apache2/apache2_util.c:247]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 25]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+    char *end = strstr(str1, "[file ");
+    if (end != NULL) {
+        strncpy(waf_detail_message, str1, end - str1);
+    }
+}
+
+/**
+ * only expose short path.
+ */
+void get_short_filename(char* waf_filename) {
```
Comment (asridharan) [inline apache2/apache2_util.c:257]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 26]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+            strcpy(tmp_filename, index+1);
+            strcpy(waf_filename, tmp_filename);
+        }
+    }
+}
+
+/**
+ * get crs type and version.
+ */
+void get_ruleset_type_version(char* waf_ruleset_info, char* waf_ruleset_type, char* waf_ruleset_version) {
```
Comment (asridharan) [inline apache2/apache2_util.c:274]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 27]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+    if (type_end != NULL) {
+        strncpy(waf_ruleset_type, type_start, type_end - type_start);
+        strcpy(waf_ruleset_version, type_end + 1);
+    }
+    else {
+        strcpy(waf_ruleset_type, type_start + 1);
+    }
+}
+
+int write_file_with_lock(apr_global_mutex_t* lock, apr_file_t* fd, char* str) {
```
Comment (asridharan) [inline apache2/apache2_util.c:292]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 28]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+
+    rc = apr_global_mutex_unlock(lock);
+    if (rc != APR_SUCCESS) {
+        return WAF_LOG_UTIL_FAILED;
+    }
+
+    return WAF_LOG_UTIL_SUCCESS;
+}
+
+char *waf_current_logtime(apr_pool_t *mp) {
```
Comment (asridharan) [inline apache2/apache2_util.c:314]: @azhao155 could you make this `static`?

File: apache2/apache2_util.c [hunk 29]
```diff
@@ -187,6 +192,192 @@ char *get_env_var(request_rec *r, char *name) {
...
+    apr_time_exp_lt(&t, apr_time_now());
+
+    apr_strftime(tstr, &len, 80, "%Y-%m-%dT%H:%M:%SZ", &t);
+    return apr_pstrdup(mp, tstr);
+}
+
+/**
+ * send all waf fields in json format to a file.
+ */
+void send_waf_log(apr_global_mutex_t* lock, apr_file_t* fd, const char* str1, const char* ip_port, const char* uri, const char* time, int mode, const char* hostname, request_rec *r) {
```
Comment (asridharan) [inline apache2/apache2_util.c:328]: @azhao155 could you make this `static`?

**Review by asridharan — COMMENTED** [review]
@azhao155 just adding a few more comments to this already merged PR.

File: apache2/apache2_util.c (hunk 23 above)
Comment (azhao155) [reply apache2/apache2_util.c:233]: will change to the static for visibility.

File: apache2/apache2_util.c (hunk 24 above)
Comment (azhao155) [reply apache2/apache2_util.c:247]: will do

File: apache2/apache2_util.c (hunk 25 above)
Comment (azhao155) [reply apache2/apache2_util.c:257]: will do

File: apache2/apache2_util.c (hunk 26 above)
Comment (azhao155) [reply apache2/apache2_util.c:274]: will do

File: apache2/apache2_util.c (hunk 27 above)
Comment (azhao155) [reply apache2/apache2_util.c:292]: will  do

File: apache2/apache2_util.c (hunk 29 above)
Comment (azhao155) [reply apache2/apache2_util.c:328]: will do

File: apache2/apache2_util.c (hunk 28 above)
Comment (azhao155) [reply apache2/apache2_util.c:314]: will do

**Comment by azhao155** [comment]
add changes to static function

**Pull request merged**